core/lib/zksync_core/src/genesis.rs

```diff
@@ -32,67 +36,141 @@ use zksync_utils::{be_words_to_bytes, bytecode::hash_bytecode, h256_to_u256, u25
 
 use crate::metadata_calculator::L1BatchWithLogs;
 
+#[derive(Debug, thiserror::Error)]
+pub enum GenesisError {
+    #[error("Root hash mismatched: From Config: {0:?}, Calculated {1:?}")]
+    RootHash(H256, H256),
+    #[error("Leaf indexes mismatched: From Config: {0:?}, Calculated {1:?}")]
+    LeafIndexes(u64, u64),
+    #[error("Base system contracts mismatched: From Config: {0:?}, Calculated {1:?}")]
+    BaseSystemContractsHashes(BaseSystemContractsHashes, BaseSystemContractsHashes),
+    #[error("Commitment mismatched: From Config: {0:?}, Calculated {1:?}")]
+    Commitment(H256, H256),
+    #[error("Wrong protocol version")]
+    ProtocolVersion,
+    #[error("Error: {0}")]
+    Other(#[from] anyhow::Error),
+    #[error("DB Error: {0}")]
+    DBError(#[from] SqlxError),
```

Comment:
I quickly checked and it looks like that it's only used for the DAL invocations, it's just in some places we add `context` and in some places we don't.
If that's true, then you can do the following:

1. Remove `Other`
2. Change `DBError(#[from] SqlxError)` to `DBError(anyhow::Error)` (note lack of `#[from]` -- we don't want an implicit conversion here).
3. Change all the functions in this module to return `GenesisError` and wrap DB errors to `GenesisError::DBError`.
4. For places where we still return `SqlxError` add `context`.

Reply:
I'd still keep `#[from]` otherwise it creates too much unnecessary boilerplate

Reply:
Then any other error (including non-DB ones) would turn into `DBError`, which breaks the purpose of strong error typing (which I thought you're proponent of).

Reply:
and it's mostly the problem with using anyhow in libraries, that we can't distinguish errors. 

I could try to make sure that we use only methods with sqlx error from dal crate